router/batchrouter/asyncdestinationmanager/marketo-bulk-upload/marketobulkupload.go

```diff
@@ -1,403 +1,355 @@
-
-		uploadResponse = common.AsyncUploadOutput{
-			ImportingJobIDs:     successfulJobIDs,
-			FailedJobIDs:        append(failedJobIDs, failedJobIDsTrans...),
-			FailedReason:        `Jobs flowed over the prescribed limit`,
-			ImportingParameters: importParameters,
-			ImportingCount:      len(importingJobIDs),
-			FailedCount:         len(failedJobIDs) + len(failedJobIDsTrans),
-			DestinationID:       destinationID,
-		}
-	case "400":
-		var responseStruct common.UploadStruct
-		err := json.Unmarshal(bodyBytes, &responseStruct)
-		if err != nil {
-			return common.AsyncUploadOutput{
-				FailedJobIDs:  append(failedJobIDs, importingJobIDs...),
-				FailedReason:  "BRT: Incorrect Response from Transformer:: " + err.Error(),
-				FailedCount:   len(failedJobIDs) + len(importingJobIDs),
-				DestinationID: destinationID,
+	for _, warningJob := range warningJobs {
+		// get warningJob data
+		var warningJobRow []string
+		for _, col := range b.csvHeaders {
+			if val, ok := warningJob[col]; ok {
+				warningJobRow = append(warningJobRow, val)
+			} else {
+				warningJobRow = append(warningJobRow, "")
 			}
 		}
-		eventsAbortedStat := b.statsFactory.NewTaggedStat("events_delivery_aborted", stats.CountType, map[string]string{
-			"module":   "batch_router",
-			"destType": destType,
-		})
-		abortedJobIDs, failedJobIDsTrans := extractJobStats(responseStruct.Metadata, importingJobIDs, http.StatusBadRequest)
-		errorMessageFromTransformer := gjson.GetBytes(bodyBytes, "error").String()
-		eventsAbortedStat.Count(len(abortedJobIDs))
-		uploadResponse = common.AsyncUploadOutput{
-			AbortJobIDs:   abortedJobIDs,
-			FailedJobIDs:  append(failedJobIDs, failedJobIDsTrans...),
-			AbortReason:   string(bodyBytes),
-			AbortCount:    len(importingJobIDs),
-			FailedCount:   len(failedJobIDs) + len(failedJobIDsTrans),
-			DestinationID: destinationID,
+		// get jobID from jobToDataHash
+		hash := calculateHashCode(warningJobRow)
+		warningJobId := b.dataHashToJobId[hash]
+		if warningJobId != 0 {
+			// Even if a job has warning, it is considered as a failure
+			metadata.FailedKeys = append(metadata.FailedKeys, warningJobId)
 		}
-		if errorMessageFromTransformer != "" {
-			uploadResponse.FailedReason = errorMessageFromTransformer
-		} else {
-			uploadResponse.FailedReason = `Jobs flowed over the prescribed limit`
+		warningJobReason := warningJob[MARKETO_WARNING_HEADER]
+		if warningJobReason != "" {
+			metadata.FailedReasons[warningJobId] = warningJobReason
 		}
-	default:
-		uploadResponse = common.AsyncUploadOutput{
-			FailedJobIDs:  append(failedJobIDs, importingJobIDs...),
-			FailedReason:  string(bodyBytes),
-			FailedCount:   len(failedJobIDs) + len(importingJobIDs),
-			DestinationID: destinationID,
+	}
+
+	// calculate succeeded keys
+	for _, job := range importingList {
+		if !lo.Contains(metadata.FailedKeys, job.JobID) && !lo.Contains(metadata.WarningKeys, job.JobID) {
```

Comment:
Should failed keys be an array? We can probably get rid of array use a map instead for faster lookup. We can get keys with lo.Keys ? WDYT ?

Reply:
keeping a backlog for now